PayPal/src/main/java/com/braintreepayments/api/paypal/PayPalClient.kt

```diff
@@ -0,0 +1,352 @@
+import com.braintreepayments.api.core.BraintreeRequestCodes
+import com.braintreepayments.api.core.Configuration
+import com.braintreepayments.api.core.UserCanceledException
+import com.braintreepayments.api.paypal.PayPalPaymentIntent.Companion.fromString
+import com.braintreepayments.api.sharedutils.Json
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * Used to tokenize PayPal accounts. For more information see the [documentation](https://developer.paypal.com/braintree/docs/guides/paypal/overview/android/v4)
+ */
+@Suppress("TooManyFunctions")
+class PayPalClient @VisibleForTesting internal constructor(
+    private val braintreeClient: BraintreeClient,
+    private val internalPayPalClient: PayPalInternalClient = PayPalInternalClient(braintreeClient),
+) {
+
+    /**
+     * Used for linking events from the client to server side request
+     * In the PayPal flow this will be either an EC token or a Billing Agreement token
+     */
+    private var payPalContextId: String? = null
+
+    /**
+     * True if `tokenize()` was called with a Vault request object type
+     */
+    private var isVaultRequest = false
+
+    /**
+     * Initializes a new [PayPalClient] instance
+     *
+     * @param context          an Android Context
+     * @param authorization    a Tokenization Key or Client Token used to authenticate
+     * @param appLinkReturnUrl A [Uri] containing the Android App Link website associated with
+     * your application to be used to return to your app from the PayPal
+     * payment flows.
+     */
+    constructor(
+        context: Context,
+        authorization: String,
+        appLinkReturnUrl: Uri
+    ) : this(BraintreeClient(context, authorization, null, appLinkReturnUrl))
+
+    /**
+     * Starts the PayPal payment flow by creating a [PayPalPaymentAuthRequestParams] to be
+     * used to launch the PayPal web authentication flow in
+     * [PayPalLauncher.launch].
+     *
+     * @param context       Android Context
+     * @param payPalRequest a [PayPalRequest] used to customize the request.
+     * @param callback      [PayPalPaymentAuthCallback]
+     */
+    fun createPaymentAuthRequest(
+        context: Context,
+        payPalRequest: PayPalRequest,
+        callback: PayPalPaymentAuthCallback
+    ) {
+        braintreeClient.sendAnalyticsEvent(PayPalAnalytics.TOKENIZATION_STARTED)
+        if (payPalRequest is PayPalCheckoutRequest) {
+            isVaultRequest = false
+            sendCheckoutRequest(context, payPalRequest, callback)
+        } else if (payPalRequest is PayPalVaultRequest) {
+            isVaultRequest = true
+            sendVaultRequest(context, payPalRequest, callback)
+        }
+    }
+
+    private fun sendCheckoutRequest(
+        context: Context,
+        payPalCheckoutRequest: PayPalCheckoutRequest,
+        callback: PayPalPaymentAuthCallback
+    ) {
+        braintreeClient.sendAnalyticsEvent("paypal.single-payment.selected", analyticsParams)
```

Comment:
These might have accidentally made it into v5 - I think all analytics events should be in the `PayPalAnalytics` enum now. Should these be removed?

Reply:
Good catch! Mind if I make this change in a follow-up PR? I feel we should reduce the amount of changes in the Kotlin conversion PR.

Reply:
Opened https://github.com/braintree/braintree_android/pull/1093